### ISSUE TITLE
ERROR: ftrace dump failed on rhel8

### ISSUE THREAD

**Issue by someconservator** [HewlettPackard/LinuxKI#55]
```
# cat /etc/redhat-release;uname -sr;/opt/linuxki/runki -f
Red Hat Enterprise Linux release 8.7 (Ootpa)
Linux 4.18.0-425.10.1.el8_7.x86_64
=== runki for Linux version 7.5
=== Disabling TTWU_QUEUE scheduling feature ===
=== Starting: Linux KI ftrace trace dump for 20 seconds ===
===
Command line: /opt/linuxki/kiinfo -dump dur=20,nop,nop,debug_dir=/sys/kernel/debug -ts 0310_0825

kiinfo (7.6)

Pausing 10 seconds to sync CPU clocks, please wait...
=== ERROR: ftrace dump failed ===
Restoring Tracing events
===
=== KI trace complete.  Collecting supplemental files.
=== Enabling TTWU_QUEUE scheduling feature ===
=== Running lsof ...
```

**Comment by MarkCRay** [comment]
Could you look at the ki.err.0310_0825 file and identify why kiinfo failed.   Are there any errors or did kiinfo coredump?

Meanwhile, I'll spin up an RHEL 8.7 VM and do some testing.   It worked fine on RHEL 8.6.

**Comment by MarkCRay** [comment]
I tested on RHEL 8.7 and had no issues.   Collect the ki.err file and check for errors.

**Comment by someconservator** [comment]
# head -n40 ki.err.0314_1601
=== runki for Linux version 7.5
=== /opt/linuxki/config:RUNKI_OPTS=
=== /opt/linuxki/config:RUNKI_AUX_CMDS=
=== /opt/linuxki/config:PERF_PROFILE_OPTS=
=== /opt/linuxki/config:SAR_OPTS=-dpqrwWB -u ALL -P ALL -n DEV -n EDEV -I SUM
=== trace invocation: /opt/linuxki/runki -f
=== Disabling TTWU_QUEUE scheduling feature ===
=== Starting: Linux KI ftrace trace dump for 20 seconds ===
=== Starting: kiinfo -dump dur=20,nop,nop,debug_dir=/sys/kernel/debug -ts 0314_1601
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  2: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  11: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  7: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  9: errno 19 - No such device
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  6: errno 19 - No such device
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  3: errno 19 - No such device
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  12: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  14: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  4: errno 19 - No such device
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  15: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  16: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
kiinfo error: read_raw_trace():268 [rrt.c]: Unable to read trace_pipe_raw - CPU:  5: errno 19 - No such device
failed to affinitize per-CPU thread, error=22
failed to affinitize per-CPU thread, error=22

**Comment by MarkCRay** [comment]
Thank you.   Its interesting that the open() of the trace_pipe_raw file didn't fail, but the read() failed with errno 19 (EBADF).   

How many logical CPUs are on this server?  I did find an issue with the ftrace tracing where the number of CPUs is limited to 500.   If we overflow the array, it may cause some unexpected results or coredump the kiinfo program.      I'll code a fix for testing if that is the case.

**Comment by someconservator** [comment]
```
# lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              2
On-line CPU(s) list: 0,1
Thread(s) per core:  1
Core(s) per socket:  1
Socket(s):           2
NUMA node(s):        1
Vendor ID:           GenuineIntel
BIOS Vendor ID:      GenuineIntel
CPU family:          6
Model:               45
Model name:          Intel(R) Xeon(R) CPU E5-2640 v3 @ 2.60GHz
BIOS Model name:     Intel(R) Xeon(R) CPU E5-2640 v3 @ 2.60GHz
Stepping:            2
CPU MHz:             2593.993
BogoMIPS:            5187.98
Hypervisor vendor:   VMware
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            20480K
NUMA node0 CPU(s):   0,1
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx hypervisor lahf_lm pti ssbd ibrs ibpb stibp tsc_adjust arat md_clear flush_l1d arch_capabilities
```

**Comment by MarkCRay** [comment]
Thanks for the lscpu.   Interesting that the VM only has 2 vCPUs, but all the errors about the CPUs are for CPUs > 2.    In the function setup_percpu_readers, it checks to see if the per-cpu trace_pipe_raw file exists:

        for ( i = 0; i < nrcpus; i++ ) {
                sprintf (percpu_file, "%s/tracing/per_cpu/cpu%d/trace_pipe_raw", debug_dir, i);
                if (stat(percpu_file, &statbuf) == 0) {
                /* if ( CPU_ISSET_S(i, size, mask) ) { */
                        alloc_thread_this_cpu(i);
                        if (debug) printf("CPU %d found \n",i);
                }
        }

Later, in run_reader_thread, it will open the per-cpu trace_pipe_raw and try to read it.   

Can you do an "ls /sys/kernel/debug/tracing/per_cpu/cpu*/trace_pipe_raw" and let me know what is in the directory?

Did this VM have more vCPUs and then was dynamically resized to to 2 vCPUs?   I need to understand why there are trace_pipe_raw files for CPUs that do not exist.

**Comment by someconservator** [comment]
This VM was created with 2 CPUs and always has 2 CPUs. God (or someone besides me) knows why it has these records for extra processors.

**Comment by MarkCRay** [comment]
Thanks.    I'm not sure why those CPUs are showing up here, but at least I understand the LinuxKI behavior.   I'll have to add another check in here.

**Comment by someconservator** [comment]
> For the per-cpu behavior, see also - https://access.redhat.com/solutions/6282021.

Ah! I understand now. But this variable still cannot be set on virtual machines due to the fact that the CPU is periodically added to some of them. And for heavily loaded productive systems, changing this variable can be a problem.

**Comment by MarkCRay** [comment]
I was able to duplicate the issue and modify the code to fix it.    

If you need an intermediate release, you can send me email at mark.ray@hpe.com.   

It will be fixed in the next release (LinuxKI 7.7).

**Comment by MarkCRay** [comment]
Fixed in LinuxKI 7.7, which is now available.
